File: Python/2024/8. August/Codeforces/E. Building an Aquarium.py
# DATE ==> 19 AUGUST 2024
# TIME ==> 1:58PM - 2:45PM
# PROBLEM NAME ==> E. Building an Aquarium
# PROBLEM TAGS ==> Binary Search, Sortings
# PROBLEM RATING ==> 1100
# AUTHOR ==> SUMONTO

def find_water(walls, height):
    answer = 0
    walls = [min(i,height) for i in walls]
    for i in walls:
        answer += height-i
    return answer

for test_cast in range(int(input())):
    lenght, water = list(map(int,input().split()))
    walls = list(map(int,input().split()))
    
    find_water(walls, 5)
    minimum,maximum = 0, water+max(walls)
    while(minimum <= maximum):
        middle = int(minimum/2 + maximum/2)
        if(find_water(walls, middle) <= water and find_water(walls,middle+1) > water):
            print(middle)
            break
        elif(find_water(walls, middle) < water): minimum = middle+1
        else: maximum = middle-1
    

File: C , C++/2024/8. August/CodeForces/E_Building_an_Aquarium.cpp
// DATE ==> 26 AUGUST 2024;
// AUTHOR ==> SUMONTO;
// PROBLEM NAME ==> E. Building an Aquarium;
// PROBLEM RATING ==> 1100;
// PROBLEM TAGS ==> Binary Search, Sortings;

#include <iostream>
#include <vector>
using namespace std;

long long max_element(vector<long long> a){
    long long answer = 0;
    for(int i=0; i<a.size(); i++){
        if (answer < a[i]) answer = a[i];
    }
    return answer;
}

long long find_water(vector<long long> a, long long height){
    long long  ans = 0;
    for(int i=0; i<a.size(); i++){
        if(height-a[i] > 0) ans += height - a[i];
    }
    return ans;
}

int main(){
    int tt;
    cin >> tt;
    while (tt--){
        long long n,x;
        cin >> n >> x;
        vector <long long> a(n);
        for(int i=0; i<n; ++i)
            cin >> a[i];

        // finding height;
        long long answer = -1;
        long long min = 0, max = x + max_element(a);
        while(min <= max){
            long long mid = min + (max-min)/2;
            if(find_water(a,mid) <= x){
                answer = mid;
                min = mid+1;
            }
            else max = mid-1;
        }

        cout << answer<< endl;
    }

    return 0;
}